source/SpinalHDL/Other language features/utils.rst

```diff
@@ -59,6 +59,7 @@ You can get more information about how hardware data types are managed on the :r
 
 .. note::
    If you use the ``cloneOf`` function on a ``Bundle``, this ``Bundle`` should be a ``case class`` or should override the clone function internally.
+   REVIEWME You can't say that and not provide an example clone() implementation / details of considerations with the implementation.
```

Comment:
Note to self, still need to provide this.

Reply:
I think we can: The preferred way should be to make it a `case class`. AFAIR we haven't needed to make a Bundle a `class` for `lib` - I'd not spend much space here to describe something that most users will never need a can be done automatically.
I thought about linking to the official docs here, but they are not much more helpful: https://www.scala-lang.org/api/current/scala/collection/mutable/Cloneable.html
But the link might still be better then just mentioning `clone` without the trait or anything

Reply:
Can we emit a code fragment example with this ?

In particular the `should override the clone function internally` is the part that needs to be clarified.
The example would confirm the exact interface/trait being provided and what a valid looking implementation can look like, even if this seems obvious to you, seeing it in code helps clarify to the reader it is nothing special.

```
class MyBundle() extend Bundle {
  // when not a case class there is no implicit clone() implementation ???
  val signalOne = Bool()
  val signalTwo = UInt(4 bits)
  override def clone(): this.type = {
    new MyBundle()  // ????
  }
}

val inst = cloneOf(bund)

```

Reply:
Here is an example : 

```scala
    class MyBundle(ppp : Int) extends Bundle {
      val a = UInt(ppp bits)
      override def clone = new MyBundle(ppp)
    }
    val x = new MyBundle(3)
    val typeDef = HardType(new MyBundle(3))
    val y = typeDef()

    cloneOf(x) //need clone methode, else it crash
    cloneOf(y) //Is ok
```

Note that HardType fix the issue, so, it also fix it when used for Stream / Vec / ... type parameters

Reply:
This is on preview here: https://dlmiles.github.io/SpinalDoc-RTD/docs_multi/html/latest/SpinalHDL/Other%20language%20features/utils.html#cloning-hardware-datatypes

Reply:
> If you use the cloneOf function on a Bundle, this Bundle should be a case class or should override the clone function internally.

Hmm that's not realy true, as long as your class has no parameter, that's fine :)

Reply:
Maybe it could be : 
"The cloneOf(Bundle) function can only be used on Bundle which either have no parameters, have the clone function defined or are case class."